src/System.Text.RegularExpressions/src/System/Buffers/MemoryOrPinnedSpan.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Buffers
+{
+    internal unsafe readonly struct MemoryOrPinnedSpan<T>
+    {
+        private readonly ReadOnlyMemory<T> _memory;
+        private readonly char* _pinnedSpan;
+
+        public MemoryOrPinnedSpan(ReadOnlyMemory<T> memory)
+        {
+            _memory = memory;
+            Length = memory.Length;
+            _pinnedSpan = null;
+        }
+
+        public MemoryOrPinnedSpan(char* pinnedSpan, int length)
+        {
+            _pinnedSpan = pinnedSpan;
+            Length = length;
+        }
+
+        public static MemoryOrPinnedSpan<T> Empty { get; } = new MemoryOrPinnedSpan<T>();
```

Comment:
I do not think that caching zeros in a static is profitable. This should just return the value without caching it.

Reply:
> This should just return the value without caching it.

Right, that's what `public static MemoryOrPinnedSpan<T> Empty => default;` does.

Reply:
I was caching it to avoid an extra instantiation per API call but I'm fine with just using default. It's right that we don't need the Empty field as I could just use default instead at the call site...